### ISSUE TITLE
OR-Tools integer variable changed to be binary internally in SCIP, Change not reflected by OR-Tools state

### ISSUE THREAD

**Issue by felixguendling** [google/or-tools#3600]
Thank you for providing OR-Tools! The interface is nice!

**What version of OR-Tools and what language are you using?**
Version: v9.5 (stable branch HEAD)
Language: C++

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
SCIP

**What operating system (Linux, Windows, ...) and version?**
Linux

**What did you do?**

```cpp
std::unique_ptr<gor::MPSolver> solver_{
      gor::MPSolver::CreateSolver("SCIP")};

// note: it's an integer variable, but currently 0/1, so SCIP changes the type to BINARY
auto const var = solver_->MakeIntVar(0.0, 1.0);

// doesn't do anything, or-tools believes it's a INTEGER
// (doesn't know about the internal change in SCIP)
var->SetInteger(true);

// not good! SCIP has var stored as BINARY
var->SetUB(2.0);
```

**What did you expect to see**

`MakeIntVar` should make an integer variable, independent of the provided bounds.

But even if creating a [0,1] valued integer variable leads to a BINARY variable, at least `var->SetInteger(true)` should change it back to INTEGER.

**What did you see instead?**

```
 /home/.../project/cmake-build-debug/_deps/scip-src/src/scip/var.c:7043: varProcessChgUbGlobal: Assertion `var->vartype != SCIP_VARTYPE_BINARY || SCIPsetIsEQ(set, newbound, 0.0) || SCIPsetIsEQ(set, newbound, 1.0)' failed.
```

**Anything else we should know about your project / environment**

I would like to change the model incrementally. The goal is to optimize seat reservation (in buses, trains, flights, etc.). So it would be great to solve the model incrementally (for each booking: check feasibility). That's why, in the beginning, the values can be `0 <= x <= 1` and later it changes to `0 <= x <= N`  for `N=2, 3, ..` as more and more bookings come in.

Also regarding this, there is the following code comment:
https://github.com/google/or-tools/blob/92eab40155d5566039c92662632f196174f8da47/ortools/linear_solver/scip_interface.cc#L686-L687

This matches my measurements: creating a completely new solver object vs. reusing and adapting constraints as new bookings come in does not improve the performance. This is an indicator that the solver starts from scratch instead of reusing information from previous runs.

As I understand the SCIP homepage, they advertise the C++ API to support "modifying and iterated solving":

![image](https://user-images.githubusercontent.com/1915889/209135163-717ec7b0-9bf4-4029-832f-4ca00248bf74.png)

So I guess that this feature of SCIP was not available at the time of the implementation of the `scip_interface.cc` and therefore cannot be expected to be supported by OR-Tools?

**Comment by felixguendling** [comment]
Ah, now I understood: since booleans (1/0) are already integers, what I tried to do doesn't make sense. It's just a limitation that SCIP changes the type to binary if the range is 0-1 and then I cannot change it back to `SCIP_VARTYPE_INTEGER` easily. But maybe this has to do with the fact, that OR Tools assumes that SCIP doesn't support incrementally anyway. So maybe I need to try using SCIP directly without the wrapper for now.

**Comment by Mizux** [comment]
AFAIK, incrementality will be a feature of the incoming math_opt

**Comment by felixguendling** [comment]
Thank you for your fast response! I will look into it. I'm also trying to use CP SAT now.

**Comment by Mizux** [comment]
Only available if using the Bazel based build and only available in C++.
It is already production ready BUT api/implem is still subject to heavy breaking change...

**Comment by lperron** [comment]
Closing the bug now.

Good luck.

**Comment by danbopes** [comment]
We encountered this bug as well. I'd like to reuse solver interfaces without having to call `.Reset()` on them to avoid the overhead of copying over all the variables/constraints to a new solver. Changing the variables, and updating the bounds seems like the best plan, but because of this bug, this is not possible ATM. I've tried setting a variable to `0, 10` and updating the bounds afterwards, but as soon as I `.Solve()` the first time, it will convert the variables to binary, and updating the bounds afterwards will cause an `ABNORMAL` output. I'm hoping there's something the or-tools team can implement to mitigate this.
